shell/shared/db.js

```diff
@@ -86,6 +86,46 @@ Grains = new Mongo.Collection("grains");
 //   publicId:  An id used to publicly identify this grain. Used e.g. to route incoming e-mail and
 //       web publishing. This field is initialized when first requested by the app.
 
+RoleAssignments = new Mongo.Collection("roleAssignments");
+// Edges in the permissions sharing graph.
+//
+// To share a permission with another user is to declare "if I have this permission, then so should
+// the recipient". A bundle of such declarations by a single sharer directed at a single recipient
+// for a single grain is called a "role assignment". A grain's owner always has every permission,
+// but permissions for other users must be computed from this collection.
+//
+// Every role assignment implicitly shares the special "can view grain" permission, which dictates
```

Comment:
An alternative design might be: if a user cannot open a grain, then they are no way of knowing whether the grain exists. But then what about Bob? After the Alice -> Bob RoleAssignment is revoked, is Bob allowed to revoke the Bob -> Carol RoleAssignment? If yes, then we can't garbage collect it when the grain gets deleted, because doing so would allow Bob to detect that the grain no longer exists.

In any case, I think these questions about who knows which grains exist are somewhat orthogonal to this pull request, and can be resolved later.